---
## Projectr : 하루시작 지하철 혼잡도 분석 02 
## Description : 
    - 서울 교통공사 역별 일별 시간대별 승하차인원 정보
    [source link] (https://data.seoul.go.kr/dataList/OA-12921/F/1/datasetView.do#)
    
## Author : Forrest Dpark (분석 담당)
## Date : 2024.05.31 ~
## Detail : 
## Update: 
    - 2024.06.02 pdg : 일별 데이터 정제 
        - 기존 테스트 데이터 의 일별 데이터가 없는 부분 참고하여 일별 데이터 추출 및 데이터 칼럼 정제를 시작.


In [279]:
import pandas as pd ,numpy as  np # Data manipulatioin
import seaborn as sns,matplotlib.pyplot as plt # visiulization
import warnings;warnings.filterwarnings('ignore') # 경고 무시 
# 기본 세팅
def plotSetting(pltStyle):
    import matplotlib.pyplot as plt # visiulization
    import platform
    from matplotlib import font_manager, rc # rc : 폰트 변경 모듈font_manager : 폰트 관리 모듈
    plt.style.use(pltStyle)
    plt.rcParams['axes.unicode_minus'] = False# unicode 설정
    if platform.system() == 'Darwin': rc('font', family='AppleGothic') # os가 macos
    elif platform.system() == 'Windows': # os가 windows
        path = 'c:/Windows/Fonts/malgun.ttf' 
        font_name = font_manager.FontProperties(fname=path).get_name()
        rc('font', family=font_name)
    else:
        print("Unknown System")
    print("___## OS platform 한글 세팅완료 ## ___")
# graph style seaborn
plotSetting("seaborn-v0_8")


___## OS platform 한글 세팅완료 ## ___


In [280]:
# Data loading 
subway = pd.read_csv("../Data/교통공사데이터/서울교통공사_역별 일별 시간대별 승하차인원 정보_20221231.csv", encoding='euc-kr')
subway.head()

연번        수송일자  호선 고유역번호(외부역코드)   역명 승하차구분  06시이전  06-07시간대  07-08시간대  \
0   1  2022-01-01   1          150  서울역    승차    120       137       211   
1   2  2022-01-01   1          150  서울역    하차    113       560       617   
2   3  2022-01-01   1          151   시청    승차     38        66       101   
3   4  2022-01-01   1          151   시청    하차     31       195       224   
4   5  2022-01-01   1          152   종각    승차     44        71        86   

   08-09시간대  ...  15-16시간대  16-17시간대  17-18시간대  18-19시간대  19-20시간대  20-21시간대  \
0       439  ...      1566      1686      1591      1358      1062       899   
1       910  ...      1329      1251      1126       884       764       654   
2       139  ...       474       550       672       528       420       434   
3       380  ...       408       377       354       213       131        98   
4       158  ...       889       964      1024       803       855      1099   

   21-22시간대  22-23시간대  23-24시간대  24시이후  
0      1327       814       234    NaN  
1       728       416       131    NaN  
2       491       232        38    NaN  
3       137        61        24    NaN  
4      1209       255        62    NaN  

[5 rows x 26 columns]

In [281]:
def dataInfoProcessing(df, replace_Nan=False):
    ''' 
    # Fucntion Description :  Data frame 의 정제해야할 부분을 체크해주는 함수 
    # Date : 2024.06.02 
    # Author : Forrest D Park 
    # update : 
        * 2024.06.02 by pdg: 일별 데이터 정제 
            - 데이터에 null 이 있음을 발견, data 정제 함수 update 
            - 함수에서 replace_Nan 아규 멘트 받아서 true 일경우 nan 을 0 으로 대체 하게 만듬. 
    '''
    print("\n1. Data colum numbers : ",len(df.columns))
    #print(subway.columns)
    #print(subway.info())
    null_message =f"총 {df.isnull().sum().sum()}개의 null 이 있습니다!" if df.isnull().sum().sum() else "Null 없는 clean data!"
    print("\n2. null ceck 결과",null_message)
    ### Null 이 있는 칼럼 추출
    haveNullColumn =[]
    for idx, col in enumerate(df.columns):
        if df[f"{col}"].isnull().sum():
            print(f"   => {idx}번째.[{col}]컬럼 : ",f"null {df[f'{col}'].isnull().sum()} 개,\t not null {df[f'{col}'].notnull().sum()} 개")
            ## Null data fill
    if replace_Nan : ## nan 을 0 으로 대체 
        df[col].fillna(value=0, inplace=True)  
    
    print("\n3. Column  Information (중복체크)")
    print( "\tidx.columName |\t\t\t\t |Colum Info(dtype)|** ")
    print( "\t","--"*len("columIdx |\t\t\t\t **|Col(dtype)|** "))
    for idx, col in enumerate(df.dtypes.keys()):
        if idx< 7:
            if len(f"\t{idx}.[{col}({df.dtypes[col]})]:")<20:
                print(f"\t{idx}.[{col}({df.dtypes[col]})]:",\
                    f"{len(df[col].unique())}/{len(df[col])} [uniq/raw]", sep=" \t\t\t")
            else:
                    print(f"\t{idx}.[{col}({df.dtypes[col]})]:",\
                    f"{len(df[col].unique())}/{len(df[col])} [uniq/raw]", sep=" \t\t")

    else: 
        print(f"\t ...etc (추가로 {len(df.dtypes.keys())-5}개의 칼럼이 있습니다 )")
   
              
dataInfoProcessing(subway)

# subway.info(show_counts=False)
# subway.value_counts('수송일자')






1. Data colum numbers :  26

2. null ceck 결과 총 82346개의 null 이 있습니다!
   => 25번째.[24시이후]컬럼 :  null 82346 개,	 not null 116734 개

3. Column  Information (중복체크)
	idx.columName |				 |Colum Info(dtype)|** 
	 ----------------------------------------------------------------
	0.[연번(int64)]: 			199080/199080 [uniq/raw]
	1.[수송일자(object)]: 			365/199080 [uniq/raw]
	2.[호선(int64)]: 			8/199080 [uniq/raw]
	3.[고유역번호(외부역코드)(object)]: 		558/199080 [uniq/raw]
	4.[역명(object)]: 			244/199080 [uniq/raw]
	5.[승하차구분(object)]: 			2/199080 [uniq/raw]
	6.[06시이전(int64)]: 			1596/199080 [uniq/raw]
	 ...etc (추가로 21개의 칼럼이 있습니다 )


> 25번째 칼럼인 24시 이후 데이터의 null 은 운행을 하지 않아서 null 인가?

In [282]:
interesting_colnumbers = [i for i in range(1,7)] + [-2,-1]
print(interesting_colnumbers)
subway[subway['24시이후'].isnull()].iloc[:,interesting_colnumbers].head()

[1, 2, 3, 4, 5, 6, -2, -1]


수송일자  호선 고유역번호(외부역코드)   역명 승하차구분  06시이전  23-24시간대  24시이후
0  2022-01-01   1          150  서울역    승차    120       234    NaN
1  2022-01-01   1          150  서울역    하차    113       131    NaN
2  2022-01-01   1          151   시청    승차     38        38    NaN
3  2022-01-01   1          151   시청    하차     31        24    NaN
4  2022-01-01   1          152   종각    승차     44        62    NaN

In [283]:
havenulldate = subway[subway['24시이후'].isnull()].iloc[:,a]
havenulldate['수송일자'].unique()[[1,2,-2,-1]]
# 2022 년 1 월 1일부터  5월 31 일까지 는 12시 이후 영업을 안한 모양.. -> 코로나 단축운행!! 


array(['2022-01-02', '2022-01-03', '2022-05-30', '2022-05-31'],
      dtype=object)

> Nan 을 0 으로 대체 

In [284]:

dataInfoProcessing(subway,replace_Nan=True)


1. Data colum numbers :  26

2. null ceck 결과 총 82346개의 null 이 있습니다!
   => 25번째.[24시이후]컬럼 :  null 82346 개,	 not null 116734 개

3. Column  Information (중복체크)
	idx.columName |				 |Colum Info(dtype)|** 
	 ----------------------------------------------------------------
	0.[연번(int64)]: 			199080/199080 [uniq/raw]
	1.[수송일자(object)]: 			365/199080 [uniq/raw]
	2.[호선(int64)]: 			8/199080 [uniq/raw]
	3.[고유역번호(외부역코드)(object)]: 		558/199080 [uniq/raw]
	4.[역명(object)]: 			244/199080 [uniq/raw]
	5.[승하차구분(object)]: 			2/199080 [uniq/raw]
	6.[06시이전(int64)]: 			1596/199080 [uniq/raw]
	 ...etc (추가로 21개의 칼럼이 있습니다 )


>  해당 데이터는 8개의 호선이 있고 역번호는 558개, 역명은 244 개가 있다.. 왜?  역번호와 역 명은 1:1 이어야 하는게 아닌가?

In [285]:
## 역명 조회
# 지하철 역 테이블 과 역번호 를 따로 만들어서 관리하도록 하자. csv 저장 필요 
역이름 = []
호선 = []
역코드 =[]


for i in subway['역명'].unique():
    for 순서, 역명 in enumerate(subway['역명']):
        if 역명==i :
            # print(f"{subway['호선'][순서]}호선 {i}, {subway['고유역번호(외부역코드)'][순서]}")
            # seoulSubwayInfo[f'{i}']= [f"{subway['호선'][순서]}호선", f"{subway['고유역번호(외부역코드)'][순서]}"]
            역이름.append(i)
            호선.append(f"{subway['호선'][순서]}호선")
            역코드.append(int(f"{subway['고유역번호(외부역코드)'][순서]}"))
            break

In [286]:
## subway column  csv 로 저장 
df= pd.DataFrame(
    {   
    "역코드" : 역코드,
    "호선" : 호선,
    "역이름" : 역이름,
    }, 
)
df.to_csv("../Data/SubwayInfo.csv", index=None)
df.pivot_table(
    # columns='호선'
    index='호선',
    aggfunc=np.size
)

역이름  역코드
호선           
1호선   10   10
2호선   48   48
3호선   30   30
4호선   22   22
5호선   49   49
6호선   30   30
7호선   40   40
8호선   15   15

--- 
> 역코드 unique 는 왜 500 개가 넘을까?
> 승차에는 없는 띄어쓰기가 하차에는 있음! 그리고 132480번째 데이터   에는 띄어쓰기가 들어가있음. 

---
###  마지막 수송 일자를 기준으로 역 이름과 역 코드를 판별한다. 



In [403]:


print("첫 수송일자 :",list(subway['수송일자'])[0])
print("마지막 수송일자 :",list(subway['수송일자'])[-1])

첫 수송일자 : 2022-01-01
마지막 수송일자 : 2022-12-31


In [315]:
### 역코드 obj -> int 로 변환  ** 아무것도 없는 데이터는 000 으로 변환 
new_stationCode = []
for i, code in enumerate(subway['고유역번호(외부역코드)']):
    # print(str(i).replace(" ",""))
    try : 
        new_stationCode.append(int(str(code)))
    except ValueError : 
        print(f"{i}번째 데이터 {code}" ,"<-value errer: ")
        new_stationCode.append(000)
        continue
print("대체됨" if len(new_stationCode)== len(subway)else "대체 안됨")

subway['고유역번호(외부역코드)'] = new_stationCode
len(subway['고유역번호(외부역코드)'].unique())
# subway.info()
## 승차데이터만, 그리고 마지막 수송일자를 기준으로 칼럼정제
test = subway[['역명','고유역번호(외부역코드)','승하차구분']]\
    [subway['승하차구분']=='승차']\
    [subway['수송일자']==list(subway['수송일자'])[-1]]


# list(test.index)
test=test.reset_index()
test.drop(columns='index',inplace=True)
test
# test.to_csv(f"../Data/CurrentSubwayStation_{list(subway['수송일자'])[-1]}.csv",)


대체됨


In [340]:
## 역명과 고유역번호가 일치하는가?
print("역명개수:",len(test['역명'].unique()),"고유역번호개수",len(test['고유역번호(외부역코드)'].unique()))
test['역명'].value_counts()
# type(a)
test[test['역명']=='종로3가']

역명개수: 239 고유역번호개수 272


역명  고유역번호(외부역코드) 승하차구분
3    종로3가           153    승차
70   종로3가           319    승차
143  종로3가          2535    승차

> 종로 3가 는 1호선, 3호선, 5호선 세 종류의 호선마다 역이 있다.     
> 중복문제를 피하기 위해서 역_이름코드 라는 칼럼을 하나 만들어서 primary key 로 만들자

In [402]:



# 역이름_코드 칼럼 생성하여 csv 저장 
station_code= test['고유역번호(외부역코드)'].to_numpy()
station_name = test['역명'].to_numpy()
test['역_이름코드'] = station_name_code = [f"{name}_{code}" for name,code in zip(station_name,station_code)]
test.to_csv(f"../Data/StationInfo_{list(subway['수송일자'])[-1]}.csv",)
    

# test